BedrockServer.cpp

```diff
@@ -691,7 +694,7 @@ void BedrockServer::worker(SData& args,
 BedrockServer::BedrockServer(const SData& args)
   : SQLiteServer(""), _args(args), _requestCount(0), _replicationState(SQLiteNode::SEARCHING),
     _upgradeInProgress(false), _suppressCommandPort(false), _suppressCommandPortManualOverride(false),
-    _syncNode(nullptr), _shutdownState(RUNNING), _backupOnShutdown(false), _controlPort(nullptr), _commandPort(nullptr)
+    _syncNode(nullptr), _shutdownState(RUNNING), _multiWriteEnabled(true), _backupOnShutdown(false), _controlPort(nullptr), _commandPort(nullptr)
```

Comment:
This list of mutex-controlled variables is getting long.  Every one of these will reduce our total concurrency ever so slightly -- indeed, concurrency is probably the *product* of these (not the sum), and thus every variable compounds the locks of the previous.  I'm not sure how else to do it, but in general we should try to avoid creating locks wherever possible -- is there any way to avoid it here?

Reply:
The list of concurrent variables isn't *that* long, there are four 'atomic' variables in this initializer list, but I get where you're coming from. The cool thing about `std::atomic` is that these objects don't necessarily lock.

From the documentation for `std::atomic::is_lock_free`:

> All atomic types except for std::atomic_flag may be implemented using mutexes or other locking operations, rather than using the lock-free atomic CPU instructions. Atomic types are also allowed to be sometimes lock-free, e.g. if only aligned memory accesses are naturally atomic on a given architecture, misaligned objects of the same type have to use locks.
> The C++ standard recommends (but does not require) that lock-free atomic operations are also address-free, that is, suitable for communication between processes using shared memory.

Of the four `atomic` variables we initialize here, two are `atomic<bool>` and two are `atomic<enum>` (which is to say, `atomic<int>`). All four them are lock-free, which I tested by calling `is_lock_free` on each in the `BedrockServer` constructor and printing the results.

For `int`s and `bool`s at least (and probably other simple types), we get atomicity here without locking.

Reply:
Oh wow, awesome!